server/gateway.go

```diff
@@ -2501,15 +2504,125 @@ func (s *Server) gatewayHandleSubjectNoInterest(c *client, acc *Account, accName
+	if !c.srv.gateway.enabled || !subjectStartsWithGatewayReplyPrefix(c.pa.subject) {
+		return false
+	}
+	// Save original subject (in case we have to forward)
+	orgSubject := c.pa.subject
+
+	clusterPrefix := c.pa.subject[gwClusterOffset : gwClusterOffset+gwHashLen]
+	srvHash := c.pa.subject[gwServerOffset : gwServerOffset+gwHashLen]
+	subject := c.pa.subject[gwSubjectOffset:]
+
+	// Check if this reply is intended for our cluster.
+	if !bytes.Equal(clusterPrefix, c.srv.gateway.getClusterHash()) {
+		// We could report, for now, just drop.
+		return true
+	}
+
+	var route *client
+
+	// If the origin is not this server, get the route this should be sent to.
+	if c.kind == GATEWAY && !bytes.Equal(srvHash, c.srv.hash) {
+		route = c.srv.getRouteByHash(srvHash)
+		// This will be possibly nil, and in this case we will try to process
+		// the interest from this server.
+	}
+
+	// Adjust the subject
+	c.pa.subject = subject
+
+	// Use a stack buffer to rewrite c.pa.cache since we only need it for
+	// getAccAndResultFromCache()
+	var _pacache [256]byte
+	pacache := _pacache[:0]
+	pacache = append(pacache, c.pa.account...)
+	pacache = append(pacache, ' ')
+	pacache = append(pacache, c.pa.subject...)
+	c.pa.pacache = pacache
+
+	acc, r := c.getAccAndResultFromCache()
+	if acc == nil {
+		typeConn := "routed"
+		if c.kind == GATEWAY {
+			typeConn = "gateway"
+		}
+		c.Debugf("Unknown account %q for %s message on subject: %q", c.pa.account, typeConn, c.pa.subject)
+		if c.kind == GATEWAY {
+			c.srv.gatewayHandleAccountNoInterest(c, c.pa.account)
+		}
+		return true
+	}
+	// If route is nil, we will process the incoming message locally.
+	if route == nil {
+		// Check if this is a service reply subject (_R_)
+		if acc.imports.services != nil && isServiceReply(c.pa.subject) {
+			// This will map the _R_ back to a real subject and get
+			// the interest for that subject and process the message.
+			c.checkForImportServices(acc, msg)
+			return true
+		}
+		var queues [][]byte
+		if len(r.psubs)+len(r.qsubs) > 0 {
+			flags := pmrCollectQueueNames | pmrIgnoreEmptyQueueFilter
+			// If this message came from a ROUTE, allow to pick queue subs
+			// only if the message was directly sent by the "gateway" server
+			// in our cluster that received it.
+			if c.kind == ROUTER {
+				flags |= pmrAllowSendFromRouteToRoute
+			}
+			queues = c.processMsgResults(acc, r, msg, c.pa.subject, c.pa.reply, flags)
+		}
+		// Since this was a reply that made it to the origin cluster,
+		// we now need to send the message with the real subject to
+		// gateways in case they have interest on that reply subject.
+		c.sendMsgToGateways(acc, msg, c.pa.subject, c.pa.reply, queues)
+	} else if c.kind == GATEWAY {
+		// Only if we are a gateway connection should we try to route
+		// to the server where the request originated.
+		header := []byte(fmt.Sprintf("RMSG %s %s %s %v", acc.Name, orgSubject, c.pa.reply, len(msg)-LEN_CR_LF))
```

Comment:
This fast path?

Reply:
Will optimize sorry. I had that very early on and forgot to improve this.

Reply:
No worries.

Reply:
Fixed.